File: leetcode/200_Number_of_Islands.java
class Solution {
    static int[][] dir = { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };

    void dfs(char[][] grid, boolean[][] visited, int x, int y) {
        if (x < 0 || x >= grid.length || y < 0 || y >= grid[0].length) {
            return;
        }

        if (grid[x][y] == '0' || visited[x][y]) {
            return;
        }

        visited[x][y] = true;
        for (int i = 0; i < Solution.dir.length; i++) {
            int tx = x + dir[i][0];
            int ty = y + dir[i][1];
            dfs(grid, visited, tx, ty);
        }
    }

    public int numIslands(char[][] grid) {
        // DFS, Flood Fill - Time: O(n^2), Space: O(n^2)
        int n = grid.length;
        int m = grid[0].length;
        boolean visited[][] = new boolean[n][m];

        int res = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    dfs(grid, visited, i, j);
                    res += 1;
                }
            }
        }
        return res;
    }
}

File: leetcode/02/200_Number_of_Islands.py
from typing import List


class Solution:
    dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    def __init__(self):
        self.grid = None
        self.n = None
        self.m = None

    def dfs(self, x, y):
        self.grid[x][y] = None
        for dx, dy in self.dirs:
            tx, ty = x + dx, y + dy
            if 0 <= tx < self.n and 0 <= ty < self.m and self.grid[tx][ty] == '1':
                self.dfs(tx, ty)

    def numIslands(self, grid: List[List[str]]) -> int:
        # Time: O(n^2 + 2*n^2) = O(n^2), Space: O(n^2) - stack size
        self.grid = grid
        self.n, self.m = len(grid), len(grid[0])
        res = 0
        for i in range(self.n):
            for j in range(self.m):
                if self.grid[i][j] == '1':
                    self.dfs(i, j)
                    res += 1
        return res
